hazelcast/src/main/java/com/hazelcast/sql/impl/QueryResultProducer.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl;
+
+import com.hazelcast.sql.HazelcastSqlException;
+import com.hazelcast.sql.impl.row.Row;
+
+import java.util.Iterator;
+
+/**
+ * Generic interface which produces iterator over results which are then delivered to users.
+ * Returned iterator must provide rows which were not returned yet.
+ */
+public interface QueryResultProducer {
+    /**
+     * Get iterator over results. Subsequent calls must return the same instance.
```

Comment:
> Subsequent calls must return the same instance

Looks like `QueryResultProducer` is itself an iterator.

Reply:
Yes. There are no strong indications for an inheritance, so the composition is used instead because it simplifies testing.